pkg/logger/json.go

```diff
@@ -0,0 +1,283 @@
+			case string:
+				key = st
+			default:
+				key = fmt.Sprintf("%s", st)
+			}
+			vals[key] = val
+		}
+	}
+
+	err := json.NewEncoder(l.writer).Encode(vals)
+	if err != nil {
+		if _, ok := err.(*json.UnsupportedTypeError); ok {
+			vals := l.getVals(t, name, level, msg)
+			vals["warn"] = errInvalidTypeMsg
+
+			json.NewEncoder(l.writer).Encode(vals)
+		}
+	}
+}
+
+func (l *JsonLogger) getVals(t time.Time, name string, level LogLevel, msg string) map[string]interface{} {
+	vals := map[string]interface{}{
+		"message":   msg,
+		"timestamp": t.Format(l.timeFormat),
+	}
+
+	if name != "" {
+		vals["module"] = name
+	}
+
+	var levelStr string
+	switch level {
+	case LogError:
+		levelStr = "error"
+	case LogWarn:
+		levelStr = "warn"
+	case LogInfo:
+		levelStr = "info"
+	case LogDebug:
+		levelStr = "debug"
+	default:
+		levelStr = "all"
+	}
+
+	vals["level"] = levelStr
+
+	if pc, file, line, ok := runtime.Caller(defaultCallerOffset + 1); ok {
+		vals["caller"] = fmt.Sprintf("%s:%d", file, line)
+		if ok {
+			f := runtime.FuncForPC(pc)
+			vals["component"] = f.Name()
+		}
+	}
+
+	return vals
+}
+
+// Debugf prints the message and args at DEBUG level
+func (l *JsonLogger) Debug(msg string, args ...interface{}) {
+	l.log(l.Name(), LogDebug, msg, args...)
+}
+
+// Infof prints the message and args at INFO level
+func (l *JsonLogger) Info(msg string, args ...interface{}) {
+	l.log(l.Name(), LogInfo, msg, args...)
+}
+
+// Warningf prints the message and args at WARN level
+func (l *JsonLogger) Warning(msg string, args ...interface{}) {
+	l.log(l.Name(), LogWarn, msg, args...)
+}
+
+// Errorf prints the message and args at ERROR level
+func (l *JsonLogger) Error(msg string, args ...interface{}) {
+	l.log(l.Name(), LogError, msg, args...)
+}
+
+// Debugf prints the message and args at DEBUG level
+func (l *JsonLogger) Debugf(msg string, args ...interface{}) {
+	l.logWithFmt(l.Name(), LogDebug, msg, args...)
+}
+
+// Infof prints the message and args at INFO level
+func (l *JsonLogger) Infof(msg string, args ...interface{}) {
+	l.logWithFmt(l.Name(), LogInfo, msg, args...)
+}
+
+// Warningf prints the message and args at WARN level
+func (l *JsonLogger) Warningf(msg string, args ...interface{}) {
+	l.logWithFmt(l.Name(), LogWarn, msg, args...)
+}
+
+// Errorf prints the message and args at ERROR level
+func (l *JsonLogger) Errorf(msg string, args ...interface{}) {
+	l.logWithFmt(l.Name(), LogError, msg, args...)
+}
+
+// Update the logging level
+func (l *JsonLogger) SetLogLevel(level LogLevel) {
+	atomic.StoreInt32(&l.level, int32(level))
+}
+
+// Name returns the loggers name
+func (i *JsonLogger) Name() string {
+	return i.name
+}
+
+// Close the logger
+func (i *JsonLogger) Close() error {
+	return i.writer.Close()
+}
+
+type writer struct {
+	b bytes.Buffer
+	w io.Writer
+}
+
+func newWriter(w io.Writer) *writer {
+	return &writer{w: w}
+}
+
+func (w *writer) Flush() (err error) {
+	var unwritten = w.b.Bytes()
+	_, err = w.w.Write(unwritten)
```

Comment:
using the same naming for the internal element might lead to infinite loops when making a mistake i.e. I'd recommend assigning a different name either for the outer or inner object so to avoid `w.w.Write(...)`

Reply:
ack